pkg/operator/etcdendpointscontroller/etcdendpointscontroller.go

```diff
@@ -0,0 +1,171 @@
+// IP addresses for etcd. It should never depend on DNS directly or transitively.
+type EtcdEndpointsController struct {
+	operatorClient  v1helpers.OperatorClient
+	nodeLister      corev1listers.NodeLister
+	configmapLister corev1listers.ConfigMapLister
+	configmapClient corev1client.ConfigMapsGetter
+}
+
+func NewEtcdEndpointsController(
+	operatorClient v1helpers.OperatorClient,
+	eventRecorder events.Recorder,
+	kubeClient kubernetes.Interface,
+	kubeInformers operatorv1helpers.KubeInformersForNamespaces,
+) factory.Controller {
+	kubeInformersForTargetNamespace := kubeInformers.InformersFor(operatorclient.TargetNamespace)
+	configmapsInformer := kubeInformersForTargetNamespace.Core().V1().ConfigMaps()
+	kubeInformersForCluster := kubeInformers.InformersFor("")
+	nodeInformer := kubeInformersForCluster.Core().V1().Nodes()
+
+	c := &EtcdEndpointsController{
+		operatorClient:  operatorClient,
+		nodeLister:      nodeInformer.Lister(),
+		configmapLister: configmapsInformer.Lister(),
+		configmapClient: kubeClient.CoreV1(),
+	}
+	return factory.New().ResyncEvery(time.Minute).WithInformers(
+		operatorClient.Informer(),
+		configmapsInformer.Informer(),
+		nodeInformer.Informer(),
+	).WithSync(c.sync).ToController("EtcdEndpointsController", eventRecorder.WithComponentSuffix("etcd-endpoints-controller"))
+}
+
+func (c *EtcdEndpointsController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	err := c.syncConfigMap(ctx, syncCtx.Recorder())
+
+	if err != nil {
+		_, _, updateErr := v1helpers.UpdateStatus(c.operatorClient, v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+			Type:    "EtcdEndpointsDegraded",
+			Status:  operatorv1.ConditionTrue,
+			Reason:  "ErrorUpdatingEtcdEndpoints",
+			Message: err.Error(),
+		}))
+		if updateErr != nil {
+			syncCtx.Recorder().Warning("EtcdEndpointsErrorUpdatingStatus", updateErr.Error())
+		}
+		return err
+	}
+
+	_, _, updateErr := v1helpers.UpdateStatus(c.operatorClient, v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+		Type:   "EtcdEndpointsDegraded",
+		Status: operatorv1.ConditionFalse,
+		Reason: "EtcdEndpointsUpdated",
+	}))
+	if updateErr != nil {
+		syncCtx.Recorder().Warning("EtcdEndpointsErrorUpdatingStatus", updateErr.Error())
+		return updateErr
+	}
+	return nil
+}
+
+func (c *EtcdEndpointsController) syncConfigMap(ctx context.Context, recorder events.Recorder) error {
+	required := configMapAsset()
+
+	// create endpoint addresses for each node
+	nodes, err := c.nodeLister.List(labels.Set{"node-role.kubernetes.io/master": ""}.AsSelector())
+	if err != nil {
+		return fmt.Errorf("unable to list expected etcd member nodes: %v", err)
+	}
+	endpointAddresses := map[string]string{}
+	for _, node := range nodes {
+		var nodeInternalIP string
+		for _, nodeAddress := range node.Status.Addresses {
+			if nodeAddress.Type == corev1.NodeInternalIP {
+				nodeInternalIP = nodeAddress.Address
+				break
+			}
+		}
+		if len(nodeInternalIP) == 0 {
```

Comment:
had a thought.  Are ipv6 addresses configmap compliant?  If not, create the list, sort, then assign them to ordered keys that are compliant.  We don't want people to use this as an API.  we already have confused questions about "the dns names not resolving"

Reply:
Good catch... looks like v6 addr would be a problem.

https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#configmap-v1-core
>Each key must consist of alphanumeric characters, '-', '_' or '.'.

Reply:
@hexfusion @retroflexer @deads2k What should the keys be if not the IP addresses? They could be encoded (base64) values.

Reply:
I've base64-encoded the keys.

Reply:
I see you already used base64,  but if you want to see what other possible encoding methods are used, you may read here: https://link.springer.com/article/10.1186/2192-113X-1-3

Reply:
Interesting link, thanks. I have no attachment to base64 — it certainly works and is easily available in the stdlib. Happy to change it to something else if somebody has a nicer specific idea.